src/Lifecycle.js

```diff
@@ -462,10 +470,28 @@ async function _doSetLoggedIn(credentials, clearStorage) {
 
     Analytics.setLoggedIn(credentials.guest, credentials.homeserverUrl);
 
+    MatrixClientPeg.replaceUsingCreds(credentials);
+    const client = MatrixClientPeg.get();
+
+    if (credentials.freshLogin) {
+        // If we just logged in, try to rehydrate a device instead of using a
+        // new device.  If it succeeds, we'll get a new device ID, so make sure
+        // we persist that ID to localStorage
+        const newDeviceId = await client.rehydrateDevice();
```

Comment:
Is there a way to know that device ID came from rehydration? That would allow us to know the same information as `freshLogin` without needing an additional flag to track it.

Reply:
Sorry, I don't really understand this comment.  The `freshLogin` flag tells us whether we just logged in or not.  And if so, then it tries to rehydrate the device.  So at the point where the `freshLogin` flag is set, we haven't tried to rehydrate yet.  We only want to try to rehydrate if we just logged in so that we don't replace a device that is already in use with the dehydrated device.

Reply:
I guess I am trying to ask if there is any other way to know whether a rehydration is needed, instead of adding a new flag `freshLogin`. It would be nice to skip this work if it is not needed, for example.

Reply:
I see.  There's two different issues here.  The first is whether there's a dehydrated device ready to be rehydrated.  If there isn't a dehydrated device, the `rehydrateDevice` function will basically only make one call to the server and then return, so it shouldn't be much work.

The second issue is whether we just logged in, which is what the `freshLogin` flag checks for.  If, for example, someone reloads Element, we don't want it to try to dehydrate a device even if there is one available.  We want it to keep using the device it was already using.  So we only check for a dehydrated device if the user just logged in.

Reply:
Okay, I think I am following. As an alternative to storing `mx_fresh_login` in local storage when it **is** fresh, could we instead rely on knowing that the `_restoreFromLocalStorage` means we **must not** be a fresh login, and so it could pass something `freshLogin: false` through to `_doSetLoggedIn` without needing extra storage...?

Sorry if this feels a bit pedantic... 😅 I'm trying to finding chances to reduce complexity (even if a tiny amount), since login is already very complex.

Reply:
> could we instead rely on knowing that the `_restoreFromLocalStorage` means we **must not** be a fresh login

Unfortunately, we can't do that.  When we do an SSO login, it saves the state to storage, and then reloads the page (why?  I  have no idea), so that means that it will `_restoreFromLocalStorage` even though it's a fresh login.